kotlin/authenticate-with-oauth/src/main/java/com/esri/arcgisruntime/sample/authenticatewithoauth/MainActivity.kt

```diff
@@ -0,0 +1,348 @@
+  }
+
+  // define configuration for OAuth Portal using custom redirect URL to receive code after auth has been granted
+  private val oAuthConfig: OAuthConfiguration by lazy {
+    OAuthConfiguration(
+      getString(R.string.portal_url),
+      getString(R.string.oauth_client_id),
+      "${BuildConfig.APPLICATION_ID}://${getString(R.string.oauth_redirect_host)}"
+    )
+  }
+
+  // instances of Portal and PortalItem to define what is displayed in the map
+  private val portal: Portal by lazy { Portal(getString(R.string.portal_url)) }
+
+  private val portalItem: PortalItem by lazy { PortalItem(portal, getString(R.string.webmap_world_traffic_id)) }
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_main)
+
+    // setup AuthenticationManager to handle auth challenges
+    AuthenticationManager.setAuthenticationChallengeHandler(this)
+    AuthenticationManager.addOAuthConfiguration(oAuthConfig)
+  }
+
+  override fun onResume() {
+    super.onResume()
+    mapView.resume()
+
+    handleIntent(intent)
+
+    mapView.map = ArcGISMap(portalItem)
+  }
+
+  override fun onPause() {
+    mapView.pause()
+    super.onPause()
+  }
+
+  override fun onDestroy() {
+    mapView.dispose()
+    super.onDestroy()
+  }
+
+  /**
+   * Attempt to handle the Intent received by the Activity
+   * If the Intent contains an authorization code, store this in the SharedPreferences
+   *
+   * @param intent the Intent to handle
+   */
+  private fun handleIntent(intent: Intent?) {
+    intent?.authCode?.let { code ->
+      sharedPreferences.putAuthCode(code)
+    }
+  }
+
+  /**
+   * Function for handling authentication challenges.
+   *
+   * Tries to use an access token if one is stored in SharedPreferences, if the token has expired, clear the token and
+   * expiry time and begin OAuth flow.
+   *
+   * If an authorization code exists in the SharedPreferences, it's likely that the user is partially through the OAuth flow
+   * and we need to try to obtain the token by performing a OAuthTokenCredentialRequest.
+   *
+   * If there is neither an access token or an authorization code in the SharedPreferences, it's likely that this is the
+   * user's first attempt at OAuth. So begin OAuth flow.
+   *
+   * @param authenticationChallenge the authentication challenge to handle
+   * @return the AuthenticationChallengeResponse indicating which action to take
+   */
+  override fun handleChallenge(authenticationChallenge: AuthenticationChallenge?): AuthenticationChallengeResponse {
+    authenticationChallenge?.let { authChallenge ->
+
+      try {
+        // if SharedPreferences has an access token
+        sharedPreferences.accessToken?.let { accessToken ->
+          // USER_CREDENTIAL_CHALLENGE is issued when attempt to use access token fails
+          if (authChallenge.type == AuthenticationChallenge.Type.USER_CREDENTIAL_CHALLENGE) {
```

Comment:
This should not be happening if the the Portal is secured by OAuth we should not be getting a `USER_CREDENTIAL_CHALLENGE` 
@LDuncAndroid can you share the repro scenario ?

Reply:
It is issued when the access token is invalid/expired.